### PR TITLE
feat: implement unification for associated types

### PR DESCRIPTION
**Pull request by mlutze**
related to #5647

### CODE REVIEW

**Comment by mlutze** [comment]
@magnus-madsen I'm putting the associated type environment in the unification functions. This will mean a lot of passing this map around. Any concerns about that?

**Comment by magnus-madsen** [comment]
> 

Can you explain more?

**Comment by mlutze** [comment]
All the `unifyM` functions will now take an additional parameter `atenv` which maps associated type symbol to a list of implementations of that associated type. So everywhere unification is done, we will need to pass around this environment.

**Comment by magnus-madsen** [comment]
> All the `unifyM` functions will now take an additional parameter `atenv` which maps associated type symbol to a list of implementations of that associated type. So everywhere unification is done, we will need to pass around this environment.

Can we talk tomorrow offline?

**Comment by mlutze** [comment]
@magnus-madsen RFR

This adds the deferred unifications to the infer monad. Remaining important steps are:
* add syntax for equality constraints
* implement the entailment check

File: main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala [hunk 1]
```diff
@@ -27,16 +27,16 @@ object Unification {
   /**
     * Unify the two type variables `x` and `y`.
     */
-  private def unifyVars(x: Type.Var, y: Type.Var, renv: RigidityEnv)(implicit flix: Flix): Result[Substitution, UnificationError] = {
+  private def unifyVars(x: Type.Var, y: Type.Var, renv: RigidityEnv)(implicit flix: Flix): Result[(Substitution, List[(Type, Type)]), UnificationError] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:30]: Perhaps we want something richer than (Type, Type). I am almost positive want that!

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala:30]: I will let you do that in the next PR.

**Pull request merged**